[PAPILIO-17] User Profile Customization

DrakeHunterK7 opened the pull request:
# General info
Complete user profile customization (user bio, user password, user phone and user profile pic)

**Issue number**: PAPILIO-17

**Task description**: 
 - Added user profile customization for user bio, user pfp, phone number and password. 
- Improved the visuals of the user profile page
- General QoL improvements to the user profile code
- Further adherence to the MVVM pattern

# Testing
To test this feature you have to:

1. Clone the PAPILIO Backend Repo, then run the server by running the command "npm run dev" (also make sure the database url is in your .env file)
2. Open the android project in Android Studio and set up your chosen device (either emulator or actual phone connected to your PC)
3. Follow the steps in the Readme.MD found in the android project root folder
4. Start the app and create a new account
5. Navigate to the user profile page via the nav bar at the bottom
6. Press the "Edit Profile" and notice that some important UI elements pop up
7. Edit your bio
8. Press "Save Changes" and wait for your bio to be updated
9. Go back to the account menu page and come back to the user profile page and see if your new bio is still there. (Good if it is!)
10. Now press the "Edit Profile" button again and edit the user profile pic
11.  After the following the on-screen steps, verify that the user profile page now has your chosen picture as the pfp (Good if it does)
12. Repeat steps 10 and 11 but edit your password and phone number instead of the user pfp
13. Wait for success messages for both password and phone number editing 
14. Now log in as a Google account and repeat steps 5 onwards (You won't be able to edit passwords from Google accounts but everything else should be the same)

# Notes
- Unit tests and UI tests remaining (will be added tomorrow)

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/services/network/UserApi.kt:
```diff
@@ -49,6 +50,12 @@ interface IUserApiService
+    @PUT("updateUserProfile")
+    suspend fun updateUser(
+        @Body user : UserUpdate,
+
```
Remove this extra line and comma at the end of the line

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -21,12 +43,20 @@ class UserProfileActivity : AppCompatActivity()
         setContentView(view)
 
+        val userRepository : IUserRepository =
+            UserRepository(FirebaseAuth.getInstance(), userService = UserApi.retrofitService)
```
You should never instantiate a repository in a view since it is a breach in our architecture. You should instead create a view model where we pass a factory where the repository is created.

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -21,12 +43,20 @@ class UserProfileActivity : AppCompatActivity()
+        userProfileViewModel.getUserByFirebaseId()
+
         // Get the last signed in Google account so we can check if the user logged in with Google or through normal log in
         val acct = GoogleSignIn.getLastSignedInAccount(this)
```
The firebase object has an attribute telling you how the user signed in (i.e. password, google, facebook, etc.). I think you should use that instead.

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -21,12 +43,20 @@ class UserProfileActivity : AppCompatActivity()
         // If the Google account exists (which means the user used Google to log in / sign up
         if(acct != null)
         {
+            isGoogleAccount = true
```
This variable seems unused so I think you can remove it

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -47,6 +77,16 @@ class UserProfileActivity : AppCompatActivity()
             // Also display the full name and email from the user's Google account on the profile
             binding.userProfileName.setText(acct.displayName)
```
Use the firebase object instead to access user attributes & use the property access syntax on Kotlin for getter/setter methods

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -47,6 +77,16 @@ class UserProfileActivity : AppCompatActivity()
+            userProfileViewModel.userObject.observe(this, Observer {
+                binding.userProfileBio.setText(userProfileViewModel.userObject.value?.userObject?.bio)
+                binding.userProfileBioEdit.setText(userProfileViewModel.userObject.value?.userObject?.bio)
+                binding.userProfilePhone.setText("Phone: +"
```
Use property access syntax instead of 'setText'

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -55,11 +95,29 @@ class UserProfileActivity : AppCompatActivity()
+                    var countryCode = userProfileViewModel.userObject.value?.userObject?.countryCode
+                    var phone = userProfileViewModel.userObject.value?.userObject?.phone
+
+                    if(phone != null)
```
Apply formatting

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -71,6 +129,270 @@ class UserProfileActivity : AppCompatActivity()
+            }
+            else { // if we are already in editing mode, then save the changes made by the user to their profile, and call the viewmodel function that sends this info through a PUT request
+
+                val error : String? = null
```
Unused variable

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -71,6 +129,270 @@ class UserProfileActivity : AppCompatActivity()
+                var closeDialog = false
+
+
+                var validationError : String? = null
```
This is redundant since you make the type of the variable nullable and you set its initial value to null. Remove the '= null' part

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -71,6 +129,270 @@ class UserProfileActivity : AppCompatActivity()
+        }
+
+        binding.userProfileAddPhoneButton.setOnClickListener{
+            val builder = AlertDialog.Builder(this)
```
You are repeating this entire dialog creation code, maybe you can create a function that takes parameters such as title 
and a list of input fields to reuse code

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -71,6 +129,270 @@ class UserProfileActivity : AppCompatActivity()
+            passwordInput1.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
+            passwordInput1.hint = "Enter Current Password"
+
+            val passwordInput2 = EditText(this)
```
passwordInput1, passwordInput2 & passwordInput3 are not good variable names

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -71,6 +129,270 @@ class UserProfileActivity : AppCompatActivity()
+
+                var closeDialog = false
+
+                val validationError1 = userProfileViewModel.validateCountryCode(countryCodeInput.text.toString())
```
Variable names need to be better than just validationError1 or 2

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -71,6 +129,270 @@ class UserProfileActivity : AppCompatActivity()
+                .circleCrop()
+                .into(binding.userProfilePicture)
+        } else if (resultCode == ImagePicker.RESULT_ERROR) {
+            Toast.makeText(this, ImagePicker.getError(data), Toast.LENGTH_SHORT).show()
```
I would rather use a snackbar instead of a Toast. I think it looks better

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -83,4 +405,10 @@ class UserProfileActivity : AppCompatActivity()
         return super.onOptionsItemSelected(item)
     }
+
+    fun EditText.hideKeyboard() {
```
Is this function necessary? It seems unused

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/viewModel/UserProfileViewModel.kt:
```diff
@@ -0,0 +1,133 @@
+        if(editedFields.isEmpty()) return
+
+        viewModelScope.launch {
+            val userResponse = userRepository.updateUser(editedFields)
```
Unused variable

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/viewModel/UserProfileViewModel.kt:
```diff
@@ -0,0 +1,133 @@
+        return null
+    }
+
+    fun validatePassword(password : String) : String?
```
Maybe we should put these validation functions in their own file rather than inside view models since they keep getting reused

AC159 commented on app/src/main/res/layout/activity_user_profile.xml:
```diff
@@ -50,7 +50,24 @@
+                    android:layout_gravity="center_horizontal"
+                    android:text="Edit Profile"
+                    android:textStyle="bold"
+                    android:textColor="#52A89F"
```
Put colors in the correct file

AC159 commented on app/src/main/res/layout/activity_user_profile.xml:
```diff
@@ -191,10 +219,62 @@
+                            android:textSize="11sp"
+                            android:text="Add Phone Number"
+                            android:layout_gravity="center_horizontal"
+                            app:backgroundTint="#52A89F"
```
Colors should be in their own file

AC159 reviewed (changes requested):
- Apply code formatting
- Try to make functions to reuse and shorten code
- See comments

DrakeHunterK7 commented:
All the comments above by @AC159 have been addressed. Please let me know if there are any more changes required and I will implement them.

Key notes about my latest changes:
- I created a kotlin class called UtilityFunctions that acts as a companion object and contains useful specific and general purpose utility functions like validating different text fields. This should allow us to resuse the same functions across various Activities and Fragments without worrying about creating unneccessary references to the UtilityFunctions class
- I added a new color called "light_sea_green" to the colors.xml file. This green color is the central green color we have chosen as part of our color palette. I replaced all direct hex color references in activity_user_profile.xml with this. 
- In UserProfileActivity.kt, I created two functions to create and return a LinearLayout and a AlertDialog.Builder. This helps in cutting down duplicate code in the file when creating dialog. 


To do:
- Add Unit tests for the UserProfileVIewModel and UserProfileViewModelFactory viewmodels
- Add UI tests for the user profile page

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/utility/UtilityFunctions.kt:
```diff
@@ -0,0 +1,57 @@
+
+import java.util.regex.Pattern
+
+class UtilityFunctions {
```
This file requires code formatting

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -1,76 +1,359 @@
+    private lateinit var binding: ActivityUserProfileBinding
+    private lateinit var userProfileViewModel: UserProfileViewModel
+    private var isEditing: Boolean = false
+    private var isGoogleAccount: Boolean = false
```
This variable is not used

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -1,76 +1,359 @@
 
-        // Get the last signed in Google account so we can check if the user logged in with Google or through normal log in
-        val acct = GoogleSignIn.getLastSignedInAccount(this)
+        val userProfileFactory = UserProfileViewModelFactory()
```
This file also needs formatting

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -83,4 +366,49 @@ class UserProfileActivity : AppCompatActivity()
+        builder.setView(layout)
+        builder.setPositiveButton(
+            "Save",
+            DialogInterface.OnClickListener { dialog_: DialogInterface, _: Int ->
```
I don't know if you are developing on android studio but it sometimes suggests to remove redundant variables/statements if you hover over those warnings like this:
![image](https://user-images.githubusercontent.com/64218770/211125626-37152ae1-574a-42b7-a1ef-e6c945e0ddac.png)

I think you can remove these redundant keywords / unused variables

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -1,76 +1,359 @@
+
+                var closeDialog = false
+
+                var passwordQualityCheck: String? =
```
Use 'val' when possible instead of 'var'

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -1,76 +1,359 @@
+
+                var passwordQualityCheck: String? =
+                    UtilityFunctions.validatePassword(newPasswordInput.text.toString())
+                var passwordEqualityCheck: String? = UtilityFunctions.confirmPassword(
```
Use 'val' when possible instead of 'var'

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -1,76 +1,359 @@
+                    UtilityFunctions.validatePhoneNumber(phoneInput.text.toString())
+
+                if (countryCodeValidationCheck != null) {
+                    messageText.setText(countryCodeValidationCheck)
```
Use property access syntax like this:
![image](https://user-images.githubusercontent.com/64218770/211125719-4b8166eb-a00b-429c-909b-790fc4907aca.png)

You should check your code to use Kotlin semantics

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/viewModel/UserProfileViewModel.kt:
```diff
@@ -0,0 +1,83 @@
+
+    fun addEditedField(fieldName: String, fieldValue: kotlin.Any)
+    {
+        editedFields.put(fieldName, fieldValue);
```
We code in kotlin not in java so you should use kotlin semantics for maps like this:
![image](https://user-images.githubusercontent.com/64218770/211125793-0e9164d0-f194-485c-9a60-110f4fa88f95.png)

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/viewModel/UserProfileViewModel.kt:
```diff
@@ -0,0 +1,83 @@
+            val credential = user!!.email?.let { EmailAuthProvider.getCredential(it, oldPassword) }
+
+            if (credential != null) {
+                user!!.reauthenticate(credential)
```
There are two unnecessary non-null assertions here:
![image](https://user-images.githubusercontent.com/64218770/211125883-c3aac029-a348-4864-a7e8-0fd046baf626.png)

AC159 commented on app/src/main/res/layout/activity_user_profile.xml:
```diff
@@ -112,7 +138,7 @@
                                 android:layout_width="wrap_content"
                                 android:layout_height="wrap_content"
                                 android:text="LOCATION"
```
These text values should be extracted to the string resources file

AC159 reviewed (changes requested):
See comments, almost there

DrakeHunterK7 commented:
Just added Unit and UI tests. Also addressed new comments above. 

Important note: I moved some key generic functions from the loginViewModel to the more appropriate UtilityFunctions companion object, so it is accessible to all views. Accordingly, the tests for these functions have also been moved to the UtilityFunctionsTest class where they are more suited.

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/utility/UtilityFunctions.kt:
```diff
@@ -0,0 +1,87 @@
+    companion object{
+        fun validatePhoneNumber(number : String): String?
+        {
+            if(number.length != 10)
```
Format the code, it's the third time I mention this

AC159 commented on app/src/main/java/com/soen490chrysalis/papilio/view/UserProfileActivity.kt:
```diff
@@ -1,76 +1,363 @@
         }
 
         // Create Action Bar var so we can 1) display it with a proper title and 2) put a working back button on it
         var actionBar = supportActionBar
```
Put val instead of var, I keep repeating the same comments which makes me think you are not addressing them properly and thoroughly enough

AC159 commented on app/src/test/java/com/soen490chrysalis/papilio/viewModel/UserProfileViewModelTest.kt:
```diff
@@ -0,0 +1,126 @@
+    }
+
+
+
```
Remove these extra lines

AC159 reviewed (approved):
I think we are good to go
